gateway-service/src/main/java/org/zowe/apiml/gateway/security/service/schema/ZosmfScheme.java

```diff
@@ -96,21 +104,21 @@ public void apply(InstanceInfo instanceInfo) {
 
             Optional<AuthSource> authSourceOptional = authSourceService.getAuthSourceFromRequest();
             authSourceOptional.ifPresent(authSource -> {
+                // client cert needs to be translated to JWT in advance, so we can determine what is the source of it
+                if (authSource.getType().equals(AuthSource.AuthSourceType.CLIENT_CERT)) {
+                    authSource = new JwtAuthSource(authSourceService.getJWT(authSource));
+                }
                 // parse authentication source to detect the source (z/OSMF / Zowe)
                 AuthSource.Parsed parsedAuthSource = authSourceService.parse(authSource);
-                switch (parsedAuthSource.getOrigin()) {
-                    case ZOSMF:
-                        // token is generated by z/OSMF, fix set cookies
-                        removeCookie(context, authConfigurationProperties.getCookieProperties().getCookieName());
-                        setCookie(context, ZosmfService.TokenType.JWT.getCookieName(), (String)authSource.getRawSource());
-                        break;
-                    case ZOWE:
-                        // user use Zowe own JWT token, for communication with z/OSMF there should be LTPA token, use it
-                        final String ltpaToken = authSourceService.getLtpaToken(authSource);
-                        setCookie(context, ZosmfService.TokenType.LTPA.getCookieName(), ltpaToken);
-                        break;
-                    default:
-                        return;
+
+                if (AuthSource.Origin.ZOSMF.equals(parsedAuthSource.getOrigin())) {
```

Comment:
This part is the same here and thirty lines later, can we extract it?

Reply:
they are almost the same, but using different methods - setCookie vs createCookie and removeCookie vs cookies.remove , also applyToRequest is removed in v2 so I didn't see much value in investing much effort to extract common part